exercises/grep/grep_test.go

```diff
@@ -0,0 +1,87 @@
+package grep
+
+import (
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func createFiles(content []string) (filenames []string) {
+	// Parse fileContentData, making the list of filenames
+	// with their content.
+	var filename string
+	var f *os.File
+	for _, d := range content {
+		t := strings.TrimSpace(d)
+		if len(t) == 0 {
+			if len(filename) == 0 {
+				continue
+			}
+			// Close file
+			f.Close()
+			filenames = append(filenames, filename)
+			filename = ""
+			f = nil
+			continue
+		}
+		if strings.Contains(t, ".txt") {
+			filename = t
+			// Open file
+			var err error
+			f, err = os.Create(filename)
+			if err != nil {
+				panic(err)
+			}
+			continue
+		}
+		fields := strings.Split(t, "|")
+		if len(fields) == 3 {
+			// Write string into file with newline.
+			_, err := f.WriteString(strings.TrimRight(fields[1], " ") + "\n")
+			if err != nil {
+				panic(err)
+			}
+		}
+	}
+	if f != nil {
+		f.Close()
+		filenames = append(filenames, filename)
+	}
+	return
+}
+
+func deleteFiles(filenames []string) {
+	for _, file := range filenames {
+		os.Remove(file)
+	}
+}
+
+func TestSearch(t *testing.T) {
+	files := createFiles(fileContentData)
+	defer deleteFiles(files)
+
+	for _, tc := range testCases {
+		actual := Search(tc.pattern, tc.flags, tc.files)
+		if !reflect.DeepEqual(actual, tc.expected) {
+			t.Fatalf("FAIL: %s\nSearch for pattern %q\nexpected %v\nactual %v.",
+				tc.description, tc.pattern, tc.expected, actual)
+		}
+		t.Logf("PASS: %s", tc.description)
+	}
+}
+
+func BenchmarkSearch(b *testing.B) {
+	files := createFiles(fileContentData)
+	defer deleteFiles(files)
+
+	b.StopTimer()
```

Comment:
Do you think [b.ResetTimer()](https://golang.org/pkg/testing/#B.ResetTimer) would be better here?

Reply:
Oh, yeah. Since I need a stop and a reset, that makes sense.  Will push another commit.

Reply:
Looking at the code in testing/benchmark.go,  using ResetTimer is much better choice with my placement.   I should have had the StopTimer before the call to createFiles, but with the ResetTimer used instead of Start/Stop it will be fine.  Thanks!

Reply:
I think involving io.Writer and the std lib flag package would be an interesting Go twist to the project. It may be nice at some point to show that you can build real command line apps with Go. If so this is the obvious exercise to do it. That would probably ramp up the difficulty level because we would be introducing a few new topics all at once. It would have to have a bunch of hints to walk users through the process. We could move it down the track a bit towards the end as a bonus. `[paasio](https://github.com/exercism/go/tree/master/exercises/paasio)` uses io.Writer too...

Overall it seems to be a worthy Go-ification of the exercise. Adds something new and helpful!

Reply:
Ok.  I'll work with it some.  And we'll see how it looks.
I agree with the bump of difficulty level, and more content in hints.md for the exercise "extension".